components/raftstore-v2/src/operation/command/write/raw_write.rs

```diff
@@ -0,0 +1,484 @@
+// Copyright 2022 TiKV Project Authors. Licensed under Apache-2.0.
+
+use std::borrow::Cow;
+
+use engine_traits::{CF_DEFAULT, CF_LOCK, CF_WRITE};
+use kvproto::raft_cmdpb::{CmdType, RaftCmdRequest, RaftRequestHeader, Request};
+use protobuf::{CodedInputStream, Message, SingularPtrField};
+use tikv_util::Either;
+
+use crate::router::CmdResChannel;
+
+// MAGIC number to hint raw write is used.
+// TODO: use protobuf blob request seems better.
+const MAGIC_PREFIX: u8 = 0x00;
+
+/// We usually use `RaftCmdRequest` for read write request. But the codec is
+/// not efficient enough for simple request. `RawWrite` is introduce to make
+/// codec alloc less and fast.
+#[derive(Debug)]
+pub struct RawWriteEncoder {
+    header: SingularPtrField<RaftRequestHeader>,
+    buf: Vec<u8>,
+    channels: Vec<CmdResChannel>,
+    size_limit: usize,
+}
+
+impl RawWriteEncoder {
+    pub fn new(
+        mut req: RaftCmdRequest,
+        size_limit: usize,
+    ) -> Result<RawWriteEncoder, RaftCmdRequest> {
+        if !Self::allow_request(&req) {
+            return Err(req);
+        }
+
+        let mut buf = Vec::with_capacity(256);
+        buf.push(MAGIC_PREFIX);
+        req.get_header()
+            .write_length_delimited_to_vec(&mut buf)
+            .unwrap();
+
+        for r in req.get_requests() {
+            encode(r, &mut buf);
+        }
+        Ok(RawWriteEncoder {
+            header: req.header,
+            buf,
+            channels: vec![],
+            size_limit,
+        })
+    }
+
+    fn allow_request(req: &RaftCmdRequest) -> bool {
+        if !req.has_status_request() && !req.has_admin_request() {
+            // TODO: skip the check and make caller use `RawWrite` directly.
+            for r in req.get_requests() {
+                if r.get_cmd_type() != CmdType::Put
+                    && r.get_cmd_type() != CmdType::Delete
+                    && r.get_cmd_type() != CmdType::DeleteRange
+                {
+                    return false;
+                }
+            }
+        } else {
+            return false;
+        };
+        true
+    }
+
+    #[inline]
+    pub fn amend(&mut self, req: RaftCmdRequest) -> Result<(), RaftCmdRequest> {
+        if Self::allow_request(&req) && req.header == self.header {
+            let last_length = self.buf.len();
+            for r in req.get_requests() {
+                encode(r, &mut self.buf);
+            }
+            // The default size limit is 8 * 0.4 = 3.2MiB.
+            if self.buf.len() < self.size_limit {
+                Ok(())
+            } else {
+                self.buf.truncate(last_length);
+                Err(req)
+            }
+        } else {
+            Err(req)
+        }
+    }
+
+    #[inline]
+    pub fn data_size(&self) -> usize {
+        self.buf.len()
+    }
+
+    #[inline]
+    pub fn encode(self) -> (Vec<u8>, Vec<CmdResChannel>) {
```

Comment:
encode() is verb and here it actually means the encoded buf and its channels. 
I would suggest to separate encode into buf() and channels() two functions for better readability.

Reply:
buf and channel are always paired. So when buf is taken, its channel must also be taken.

Reply:
ok, then maybe take_encode would be more accurate.

Reply:
It's not `take_xx`. `take_xx` means take some field out of the struct. Both the struct and field are still valid. Here the struct is moved.